pkg/rules/handlers.go

```diff
@@ -0,0 +1,267 @@
+				Status:  http.StatusBadRequest,
+			},
+		)
+
+		return
+	}
+
+	rule := new(RuleData)
+	if err := json.Unmarshal(body, rule); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(w).Encode(
+			types.Response{
+				Message: "Error parsing JSON payload",
+				Status:  http.StatusBadRequest,
+			},
+		)
+
+		return
+
+	}
+
+	err = a.service.CreateRule(r.Context(), ruleID, *rule, fileName)
+
+	if err != nil {
+
+		rr := types.Response{
+			Status:  http.StatusInternalServerError,
+			Message: err.Error(),
+		}
+
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(rr)
+
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	json.NewEncoder(w).Encode(
+		types.Response{
+			Message: fmt.Sprintf("Created rule %s", ruleID),
+			Status:  http.StatusCreated,
+		},
+	)
+}
+
+func (a *API) handleUpdate(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	ruleID := chi.URLParam(r, "id")
+	ruleID = cleanStr(ruleID)
+	defer r.Body.Close()
+	body, err := io.ReadAll(r.Body)
+
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(w).Encode(
+			types.Response{
+				Message: "Error parsing request payload",
+				Status:  http.StatusBadRequest,
+			},
+		)
+
+		return
+	}
+
+	rule := new(RuleData)
+	if err := json.Unmarshal(body, rule); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(w).Encode(
+			types.Response{
+				Message: "Error parsing JSON payload",
+				Status:  http.StatusBadRequest,
+			},
+		)
+
+		return
+
+	}
+
+	err = a.service.UpdateRule(r.Context(), ruleID, *rule)
+
+	if err != nil {
+
+		rr := types.Response{
+			Status:  http.StatusInternalServerError,
+			Message: err.Error(),
+		}
+
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(rr)
+
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(
+		types.Response{
+			Message: fmt.Sprintf("Updated rule %s", ruleID),
+			Status:  http.StatusOK,
+		},
+	)
+}
+
+func (a *API) handleRemove(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	ruleID := chi.URLParam(r, "id")
+	ruleID = cleanStr(ruleID)
+
+	err := a.service.DeleteRule(r.Context(), ruleID)
+
+	if err != nil {
+
+		rr := types.Response{
+			Status:  http.StatusInternalServerError,
+			Message: err.Error(),
+		}
+
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(rr)
+
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(
+		types.Response{
+			Message: fmt.Sprintf("Deleted rule %s", ruleID),
+			Status:  http.StatusCreated,
+		},
+	)
+}
+
+func NewAPI(service ServiceInterface, logger logging.LoggerInterface) *API {
+	a := new(API)
+
+	a.service = service
+
+	a.logger = logger
+
+	return a
+}
+
+func cleanStr(s string) string {
+	return strings.Map(
+		func(r rune) rune {
+			if r == '\\' {
+				return -1
+			}
+			return r
+		},
+		s,
+	)
+}
```

Comment:
not sure I follow this?

Reply:
The examples I've seen for oathkeeper rules had IDs that included a :. Because of this I thought that some IDs that were parameters of the URL would include characters that might be escaped. I wrote this function to check for escaped characters in the Id parameters. I agree with your other point that having the ID as a url parameter is misguided, so I'll take this part out.

Reply:
This function has been taken out. Because the rule ids are also url parameters for some handlers, it's better for the ids to not have special characters.